addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/netutils/AsyncHttpClient.java

```diff
@@ -0,0 +1,184 @@
+        return CompletableFuture.supplyAsync(() -> new Parameters(address, "PUT", body, 1000), scheduler)
+                .thenApply(AsyncHttpClient::doNetwork);
+    }
+
+    /**
+     * Perform an asynchronous POST request
+     *
+     * @param address The address
+     * @param body The message body
+     * @param scheduler A scheduler for the thread creation
+     * @return The result
+     * @throws IOException Any IO exception in an error case.
+     */
+    public static CompletableFuture<Result> postAsync(String address, String body, ScheduledExecutorService scheduler) {
+        return CompletableFuture.supplyAsync(() -> new Parameters(address, "POST", body, 1000), scheduler)
+                .thenApply(AsyncHttpClient::doNetwork);
+    }
+
+    /**
+     * Perform an asynchronous GET request
+     *
+     * @param address The address
+     * @param scheduler A scheduler for the thread creation
+     * @return The result
+     * @throws IOException Any IO exception in an error case.
+     */
+    public static CompletableFuture<Result> getAsync(String address, ScheduledExecutorService scheduler) {
+        return CompletableFuture.supplyAsync(() -> new Parameters(address, "GET", "", 1000), scheduler)
+                .thenApply(AsyncHttpClient::doNetwork);
+    }
+
+    public static Result delete(String address, int timeout) throws IOException {
+        return doNetwork(new Parameters(address, "DELETE", "", timeout));
+    }
+
+    private static Result doNetwork(Parameters p) {
+        HttpURLConnection conn;
+        try {
+            conn = (HttpURLConnection) new URL(p.address).openConnection();
+        } catch (IOException e) {
+            throw new IllegalStateException(e);
+        }
+
+        try (AutoCloseable conc = () -> conn.disconnect()) {
+            conn.setRequestMethod(p.requestMethod);
+            conn.setRequestProperty("Content-Type", "application/json");
+            conn.setConnectTimeout(p.timeout);
+            conn.setReadTimeout(p.timeout);
+
+            String body = p.body;
+            if (body != null && !body.equals("")) {
+                conn.setDoOutput(true);
+                OutputStreamWriter out = new OutputStreamWriter(conn.getOutputStream());
+                out.write(body);
+                out.close();
+            }
+
+            if (conn.getResponseCode() / 100 == 2) {
+                String output = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8))
+                        .lines().collect(Collectors.joining("\n"));
+                return new Result(output, conn.getResponseCode());
+            } else {
+                String output = new BufferedReader(new InputStreamReader(conn.getErrorStream(), StandardCharsets.UTF_8))
+                        .lines().collect(Collectors.joining("\n"));
+                return new Result(output, conn.getResponseCode());
+            }
+        } catch (Exception e) {
```

Comment:
Why would you wrap any Exception in an IllegalStateException. When using a `try(...) {}` the catch is optional so that might make more sense?

Reply:
I'd like the exception to be handled by the `CompletableFuture.exceptionally`

Reply:
That will also happen if you will not catch it or am I mistaken?

Reply:
But then the method signature would need to contain the thrown Exceptions and a CompleteableFuture uses functional interfaces and they do not allow explicit exceptions.
It would fail at:
`CompletableFuture.supplyAsync(() -> ...).thenApply(AsyncHttpClient::doNetwork);`